src/ArrayInterface.jl

```diff
@@ -81,6 +81,16 @@ has_parent(::Type{T}) where {T} = _has_parent(parent_type(T), T)
 _has_parent(::Type{T}, ::Type{T}) where {T} = False()
 _has_parent(::Type{T1}, ::Type{T2}) where {T1,T2} = True()
 
+"""
+    buffer(x)
+
+Return the buffer data that `x` points to. Unlike `parent(x::AbstractArray)`, `buffer(x)`
+may not return another array.type.
```

Comment:
Hmmm, what does this mean? `typeof(x)` and `typeof(buffer(x))` must be the same type?

Reply:
If you're referring to `array.type`, it's a typo.

Reply:
It's not clear to me the terms "may not" and "another" here. Does "may not" mean "must not" or does it allow some sort of relaxation?

If `x` is a matrix, is it allowed for `buffer` to be a vector? If `x` is an OffsetArray, is it allowed for `buffer(x)` to be `Array`?

Or, as long as `buffer(x)` is a contiguous memory block whose element type is the same as `eltype(x)` it is a valid buffer?

Reply:
> Or maybe here buffer doesn't mean buffer block, maybe it means something like "real storage".

More like this. `parent` is typically used on an array to return a wrapped array (not sure of an instance where it doesn't actually). If we want to be able to refer to the raw storage of values for arrays like `Tridiagonal` this criteria needs to be relaxed. There are also arrays like `SparseMatrixCSC` where the buffer is still an array type but isn't return with `parent`.

I made the explanation for what `buffer` returns here intentionally lax because we shouldn't need to dictate how data is stored to much; but perhaps it's just vague. We may want to add suggestions in the future (e.g., whatever is returned has `getindex` defined, supports `Ptr` or `ManualMemory.PseudoPtr`), but I don't see an immediate need for it and documenting that sort of requirement might force us into a breaking change when we encounter exceptions later on.

For `OffsetArray` nothing needs to happen because it already has `parent` defined, which returns its buffer.

Reply:
I get your points now. Can we rename this method to something more accurate? Maybe `raw_data`? Ask because I feel the name is perhaps ambiguous here with the purpose of "allocating a small array to serve as a buffer".

Reply:
The reason I chose `buffer` is because we have `ManualMemory.preserve_buffer` which digs down to the deepest "buffer".
I think the return of `preserve_buffer` would be more accurately described as "raw data" because what I'm calling "buffer" here may have additional information attached to it, like another wrapper array.

> the name is perhaps ambiguous here with the purpose of "allocating a small array to serve as a buffer"

If we are looking at a method name for allocating array memory (outside of `similar), I think it would be odd to have a name that isn't derived from `malloc`/`alloc`.

All of this is clearly just my opinion so if there's a consensus for a clearly better name I'd be completely open to a rename.

Reply:
I've waited a couple days to see if anyone else wanted to chime, but I'm gonna register the latest version now. I don't anticipate changing it being a breaking change for other packages within the immediate future (unless someone is closely following this repo and integrating every new method immediately). So if people want to discuss it more I'd encourage you to create an issue.